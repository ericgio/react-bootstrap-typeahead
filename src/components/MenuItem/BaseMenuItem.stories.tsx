import React from 'react';
import { StoryFn, Meta } from '@storybook/react';

import { BaseMenuItem, BaseMenuItemProps } from './MenuItem';

export default {
  title: 'Components/MenuItem/BaseMenuItem',
  component: BaseMenuItem,
} as Meta;

const Template: StoryFn<BaseMenuItemProps> = (args) => (
  <div className="dropdown-menu show">
    <BaseMenuItem {...args} />
  </div>
);

export const Default = Template.bind({});
Default.args = {
  children: 'This is a base menu item',
};

export const Active = Template.bind({});
Active.args = {
  active: true,
  children: 'This is an active base menu item',
};

export const Disabled = Template.bind({});
Disabled.args = {
  children: 'This is a disabled base menu item',
  disabled: true,
};
