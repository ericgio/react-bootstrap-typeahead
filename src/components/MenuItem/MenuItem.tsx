import cx from 'classnames';
import PropTypes from 'prop-types';
import React, { forwardRef, HTMLAttributes, MouseEvent } from 'react';

import useItem, { UseItemProps } from '../../core/useItem';
import { optionType } from '../../propTypes';

export interface BaseMenuItemProps extends HTMLAttributes<HTMLAnchorElement> {
  active?: boolean;
  disabled?: boolean;
  href?: string;
}

export const BaseMenuItem = forwardRef<HTMLAnchorElement, BaseMenuItemProps>(
  ({ active, children, className, disabled, onClick, ...props }, ref) => {
    return (
      <a
        {...props}
        className={cx('dropdown-item', { active, disabled }, className)}
        href={props.href || '#'}
        onClick={(e: MouseEvent<HTMLAnchorElement>) => {
          e.preventDefault();
          !disabled && onClick && onClick(e);
        }}
        ref={ref}>
        {children}
      </a>
    );
  }
);

export type MenuItemProps = UseItemProps<HTMLAnchorElement>;

function MenuItem(props: MenuItemProps) {
  return <BaseMenuItem {...useItem(props)} />;
}

MenuItem.propTypes = {
  option: optionType.isRequired,
  position: PropTypes.number,
};

export default MenuItem;
