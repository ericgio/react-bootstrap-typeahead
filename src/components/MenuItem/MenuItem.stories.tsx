import React from 'react';
import { StoryFn, Meta } from '@storybook/react';

import {
  defaultContext,
  TypeaheadContext,
  TypeaheadContextType,
} from '../../core';

import MenuItem, { MenuItemProps } from './MenuItem';

export default {
  title: 'Components/MenuItem/MenuItem',
  component: MenuItem,
} as Meta;

interface Args {
  context: Partial<TypeaheadContextType>;
  props: MenuItemProps;
}

const value = {
  ...defaultContext,
  id: 'test-id',
};

const Template: StoryFn<Args> = ({ context, props }) => (
  <TypeaheadContext.Provider value={{ ...value, ...context }}>
    <div className="dropdown-menu show">
      <MenuItem {...props} />
    </div>
  </TypeaheadContext.Provider>
);

export const Default = Template.bind({});
Default.args = {
  props: {
    children: 'This is a menu item',
    label: 'test label',
    option: '',
    position: 0,
  },
};
