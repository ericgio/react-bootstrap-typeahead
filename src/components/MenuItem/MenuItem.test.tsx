import React from 'react';

import * as baseStories from './BaseMenuItem.stories';
import * as stories from './MenuItem.stories';

import {
  composeStories,
  generateSnapshots,
  screen,
  render,
  userEvent,
} from '../../tests/helpers';

const { Default: BaseDefault, Active, Disabled } = composeStories(baseStories);
const { Default } = composeStories(stories);

const option = '';

describe('<BaseMenuItem>', () => {
  generateSnapshots(baseStories);

  it('triggers an event when clicked', async () => {
    const user = userEvent.setup();
    const onClick = jest.fn();
    render(<BaseDefault onClick={onClick} />);

    const item = screen.getByRole('link');
    await user.click(item);

    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('renders a default href', () => {
    render(<BaseDefault />);
    expect(screen.getByRole('link')).toHaveAttribute('href', '#');
  });

  it('renders an href passed to the MenuItem', () => {
    const href = '/some/link';
    render(<BaseDefault href={href} />);
    expect(screen.getByRole('link')).toHaveAttribute('href', href);
  });

  it('renders an active base menu item', () => {
    render(<Active />);
    expect(screen.getByRole('link')).toHaveClass('active');
  });

  it('renders a disabled base menu item', async () => {
    const user = userEvent.setup();
    const onClick = jest.fn();
    render(<Disabled disabled onClick={onClick} />);

    const item = screen.getByRole('link');
    await user.click(item);

    expect(item).toHaveClass('disabled');
    expect(onClick).toHaveBeenCalledTimes(0);
  });
});

describe('<MenuItem>', () => {
  generateSnapshots(stories);

  it('renders a MenuItem', () => {
    render(<Default />);

    const item = screen.getByRole('option');
    expect(item).toHaveAttribute('aria-label', 'test label');
    expect(item).toHaveAttribute('aria-selected', 'false');
    expect(item).toHaveAttribute('id', 'test-id-item-0');
  });

  it('conditionally calls `onInitialItemChange`', () => {
    const onInitialItemChange = jest.fn();
    const context = { onInitialItemChange };

    const { rerender } = render(
      <Default context={context} props={{ option, position: 1 }} />
    );
    expect(onInitialItemChange).toHaveBeenCalledTimes(0);

    rerender(<Default context={context} props={{ option, position: 0 }} />);
    expect(onInitialItemChange).toHaveBeenCalledTimes(1);
  });

  it('changes the active state of the menu item', () => {
    render(<Default context={{ activeIndex: 0 }} />);

    const item = screen.getByRole('option');
    expect(item).toHaveClass('active');
    expect(item).toHaveAttribute('aria-selected', 'true');
  });

  it('sets the active state if it is the only result', () => {
    render(<Default context={{ isOnlyResult: true }} />);

    const item = screen.getByRole('option');
    expect(item).toHaveClass('active');
    expect(item).toHaveAttribute('aria-selected', 'true');
  });

  it('calls the click handlers', async () => {
    const user = userEvent.setup();
    const onClick = jest.fn();
    const onMenuItemClick = jest.fn();

    render(
      <Default
        context={{ onMenuItemClick }}
        props={{ onClick, option, position: 0 }}
      />
    );

    await user.click(screen.getByRole('option'));
    expect(onClick).toHaveBeenCalledTimes(1);
    expect(onMenuItemClick).toHaveBeenCalledTimes(1);
  });

  it('renders a disabled menu item', async () => {
    const user = userEvent.setup();
    const onClick = jest.fn();
    render(
      <Default props={{ disabled: true, onClick, option, position: 0 }} />
    );

    const item = screen.getByRole('option');
    await user.click(item);

    expect(item).toHaveClass('disabled');
    expect(onClick).toHaveBeenCalledTimes(0);
  });
});
