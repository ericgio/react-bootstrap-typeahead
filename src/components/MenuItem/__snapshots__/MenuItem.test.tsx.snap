// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BaseMenuItem> Active story renders snapshot 1`] = `
<div
  class="dropdown-menu show"
>
  <a
    class="dropdown-item active"
    href="#"
  >
    This is an active base menu item
  </a>
</div>
`;

exports[`<BaseMenuItem> Default story renders snapshot 1`] = `
<div
  class="dropdown-menu show"
>
  <a
    class="dropdown-item"
    href="#"
  >
    This is a base menu item
  </a>
</div>
`;

exports[`<BaseMenuItem> Disabled story renders snapshot 1`] = `
<div
  class="dropdown-menu show"
>
  <a
    class="dropdown-item disabled"
    href="#"
  >
    This is a disabled base menu item
  </a>
</div>
`;

exports[`<MenuItem> Default story renders snapshot 1`] = `
<div
  class="dropdown-menu show"
>
  <a
    aria-label="test label"
    aria-selected="false"
    class="dropdown-item"
    href="#"
    id="test-id-item-0"
    role="option"
  >
    This is a menu item
  </a>
</div>
`;
